internal/csi-addons/rbd/reclaimspace.go

```diff
@@ -0,0 +1,145 @@
+	req *rs.ControllerReclaimSpaceRequest) (*rs.ControllerReclaimSpaceResponse, error) {
+	volumeID := req.GetVolumeId()
+	if volumeID == "" {
+		return nil, status.Error(codes.InvalidArgument, "empty volume ID in request")
+	}
+
+	cr, err := util.NewUserCredentials(req.GetSecrets())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	defer cr.DeleteCredentials()
+
+	rbdVol, err := rbdutil.GenVolFromVolID(ctx, volumeID, cr, req.GetSecrets())
+	if err != nil {
+		return nil, status.Errorf(codes.Aborted, "failed to find volume with ID %q: %s", volumeID, err.Error())
+	}
+	defer rbdVol.Destroy()
+
+	err = rbdVol.Sparsify()
+	if err != nil {
+		// TODO: check for different error codes?
+		return nil, status.Errorf(codes.Internal, "failed to sparsify volume %q: %s", rbdVol, err.Error())
+	}
+
+	return &rs.ControllerReclaimSpaceResponse{}, nil
+}
+
+// ReclaimSpaceNodeServer struct of rbd CSI driver with supported methods
+// of CSI-addons reclaimspace controller service spec.
+type ReclaimSpaceNodeServer struct {
+	*rs.UnimplementedReclaimSpaceNodeServer
+}
+
+// NewReclaimSpaceNodeServer creates a new IdentityServer which handles the
+// Identity Service requests from the CSI-Addons specification.
+func NewReclaimSpaceNodeServer() *ReclaimSpaceNodeServer {
+	return &ReclaimSpaceNodeServer{}
+}
+
+func (rsns *ReclaimSpaceNodeServer) RegisterService(server grpc.ServiceRegistrar) {
+	rs.RegisterReclaimSpaceNodeServer(server, rsns)
+}
+
+// NodeReclaimSpace runs fstrim or blkdiscard on the path where the volume is
+// mounted or attached. When a volume with multi-node permissions is detected,
+// an error is returned to prevent potential data corruption.
+func (rsns *ReclaimSpaceNodeServer) NodeReclaimSpace(
+	ctx context.Context,
+	req *rs.NodeReclaimSpaceRequest) (*rs.NodeReclaimSpaceResponse, error) {
+	// volumeID is a required attribute, it is part of the path to run the
+	// space reducting command on
+	// nolint:ifshort // volumeID is incorrectly assumed to be used only once
+	volumeID := req.GetVolumeId()
+	if volumeID == "" {
+		return nil, status.Error(codes.InvalidArgument, "empty volume ID in request")
+	}
+
+	// path can either be the staging path on the node, or the volume path
+	// inside an application container
+	path := req.GetStagingTargetPath()
+	if path == "" {
+		path = req.GetVolumePath()
+		if path == "" {
+			return nil, status.Error(codes.InvalidArgument, "required parameter volume_path is not set")
+		}
+	} else {
+		// append the right staging location used by this CSI-driver
+		path = fmt.Sprintf("%s/globalmount/%s", path, volumeID)
+	}
+
+	// do not allow RWX block-mode volumes, danger of data corruption
+	isBlock, isMultiNode := csicommon.IsBlockMultiNode([]*csi.VolumeCapability{req.GetVolumeCapability()})
+	if isMultiNode {
+		return nil, status.Error(codes.Unimplemented, "multi-node space reclaim is not supported")
+	}
+
+	cmd := "fstrim"
+	if isBlock {
+		cmd = "blkdiscard"
+	}
```

Comment:
so fsmode volumes `fstrim` and block mode ``blkdiscard` - may be some constants like 
fsModeTrimCommand = `fstrim`
blockModeTrimCommand = `blkdiscard` would be better , instad of hard coded ones

Reply:
there is no benefit in adding constants for these, they are used only once